File: backend-fastapi/models/collections/place.py
from typing import TYPE_CHECKING, Awaitable, cast

from beanie import Delete, Insert, PydanticObjectId, after_event, before_event

from lib.types_ import ChangeEvent
from models.collections.base import (
    BaseDocument,
    CollectionEnum,
    document_registry,
)
from models.fields import place as PlaceFields

if TYPE_CHECKING:
    # Avoid Circular imports
    from models.collections.user import User


class Place(BaseDocument):
    # Fields
    title: PlaceFields.title_annot
    description: PlaceFields.description_annot
    imageUrl: PlaceFields.imageUrl_annot | None = None
    address: PlaceFields.address_annot
    location: PlaceFields.Location | None = None
    embedding: PlaceFields.embedding_annot | None = None

    # Relations
    creatorId: PydanticObjectId

    class Settings:
        name = CollectionEnum.PLACES
        indexes = ["createdAt"]

    @before_event(ChangeEvent)
    async def validate_creator_exists(self) -> None:
        Users = cast(type["User"], document_registry[CollectionEnum.USERS])
        if not await Users.find_one(Users.id == self.creatorId):
            raise ValueError("Creator does not exist")

    @after_event([Insert])
    async def add_place_to_user(self) -> None:
        Users = cast(type["User"], document_registry[CollectionEnum.USERS])
        query = Users.find_one(Users.id == self.creatorId).update(
            {"$addToSet": {"places": self.id}}
        )
        await cast(Awaitable, query)

    @before_event([Delete])
    async def remove_place_from_user(self) -> None:
        Users = cast(type["User"], document_registry[CollectionEnum.USERS])
        query = Users.find_one(Users.id == self.creatorId).update(
            {"$pull": {"places": self.id}}
        )
        await cast(Awaitable, query)


document_registry[CollectionEnum.PLACES] = Place

File: backend-express/src/models/collections/place.ts
import { model, Schema } from "mongoose";

import { HttpStatus } from "@/lib/express";
import { ApiError, CollectionEnum } from "@/types";

import { PlaceDB } from "../schemas";

// Schema creation
export const PlaceCollectionSchema = new Schema<PlaceDB>(
    {
        // Fields
        title: { type: String },
        description: { type: String },
        imageUrl: { type: String },
        address: { type: String },
        location: {
            _id: false,
            type: {
                lat: { type: Number },
                lng: { type: Number },
            },
        },
        // AI
        embedding: { type: [Number] },
        // Foreign Keys:
        creatorId: {
            type: Schema.ObjectId,
            ref: CollectionEnum.USERS,
        },
    },
    { timestamps: true }
);
PlaceCollectionSchema.index({ createdAt: 1 });

// Hooks
PlaceCollectionSchema.pre("save", async function (next) {
    // Lazy loading to avoid circular imports
    const { UserModel } = await import("./user");
    const userExists = await UserModel.exists({ _id: this.creatorId });
    if (!userExists)
        throw new ApiError(HttpStatus.BAD_REQUEST, "User does not exist");
    next();
});

PlaceCollectionSchema.post("save", async function (place, next) {
    // Lazy loading to avoid circular imports
    const { UserModel } = await import("./user");
    await UserModel.findByIdAndUpdate(place.creatorId, {
        $addToSet: { places: place._id },
    });
    next();
});

PlaceCollectionSchema.pre("deleteOne", async function (next) {
    // Lazy loading to avoid circular imports
    const { UserModel } = await import("./user");
    const place = await this.model.findOne(this.getFilter());
    await UserModel.findByIdAndUpdate(place.creatorId, {
        $pull: { places: place._id },
    });
    next();
});

// Model Creation
export const PlaceModel = model<PlaceDB>(
    CollectionEnum.PLACES,
    PlaceCollectionSchema
);

export type PlaceDocument = InstanceType<typeof PlaceModel>;
